chore: Added client anticipation documentation (#1215)
Unity-Technologies/com.unity.multiplayer.docs: docs/advanced-topics/client-anticipation.md

> ---

Reviewer: Little nitpick about the images. Not a blocker, but could make them a bit clearer. I'd try to put on top the "main" lesson" and have the other "support" client at the bottom. So for example, if the lesson is about reanticipation, I'd put the client timeline doing the reanticipation on top. Makes it easy to jump at the right spot when reading.
Not crazy important, as a reader I can figure it out, it just requires a bit more concentration.

- On `AnticipatedNetworkTransform`, the anticipated value is stored in both `gameObject.transform` and `anticipatedNetworkTransform.AnticipatedState`, both of which are **read-only** on the client, while the authoritative value is stored as `anticipatedNetworkTransform.AuthoritativeState`. If you simply update the transform's values on the client, they will be overwritten when the next server update comes in. To perform anticipation on the transform, you have to call `anticipatedNetworkTransform.AnticipateMove()`, `anticipatedNetworkTransform.AnticipateRotate()`,  `anticipatedNetworkTransform.AnticipateScale()`, or `anticipatedNetworkTransform.AnticipateState()` to update all three at once. As with `AnticipatedNetworkVariable<T>`, calling any of these on the server will update both the anticipated and authoritative values.

## StaleDataHandling

> On problem that an anticipation system needs to be able to solve is the problem of **stale data**. Stale data refers to updates from the server that represent actions that happened before your last request, and are actually going to be overwritten by that request.

Reviewer: ```suggestion
One problem that an anticipation system needs to be able to solve is the problem of **stale data**. Stale data refers to updates from the server that represent actions that happened before your last request, and are actually going to be overwritten by that request.
```

<figure>
<ImageSwitcher 
lightImageSrc="/sequence_diagrams/Anticipation/ServerAuthoritativeMultiClient.png?text=LightMode"
darkImageSrc="/sequence_diagrams/Anticipation/ServerAuthoritativeMultiClient_Dark.png?text=DarkMode"/>
</figure>

With client anticipation, we change the way this works: client 1 anticipates the change to blue so it happens immediately, and then later sees the object change to red, which, again, is fine. Client 2 also sees the object change to red immediately, but because a change to blue is already in progress, that overwrites client 2's anticipated value, causing it to flicker briefly to blue from client 1's request before changing back to red again from client 2's request.

<figure>
<ImageSwitcher 
lightImageSrc="/sequence_diagrams/Anticipation/StaleDataNoPolicy.png?text=LightMode"
darkImageSrc="/sequence_diagrams/Anticipation/StaleDataNoPolicy_Dark.png?text=DarkMode"/>
</figure>

> To address this, Netcode for GameObjects's client anticipation includes a feature called `StaleDataHandling`. Stale data is determined based on assumptions about causation - it assumes that when you make an anticipation on the client side based on player input, that at the same time, you will be sending an RPC to the server requesting it to make the same change. It uses a continuously incrementing `AnticipationCounter` to track when the server has received and responded to the batch of requests that was sent on the same frame as the variable was anticipated. If an update for a variable arrives before the server has processed that message, the anticipation system regards that data as being "stale".

Reviewer: Part of this feels like implementation details that could be put a special comment box? Just to make it easier to skim over when fast reading?

> There are two ways you can respond to stale data, which are determined by the `StaleDataHandling` value on each `AnticipatedNetworkVariable` and `AnticipatedNetworkTransform`:

Reviewer: I'd list those two ways here before jumping in the details. This way users that are fast reading can just jump to it without having to read every details.

Reviewer 2: Added both approaches and created sub-header sections for each approach.

<figure>
<ImageSwitcher 
lightImageSrc="/sequence_diagrams/Anticipation/StaleDataIgnore.png?text=LightMode"
darkImageSrc="/sequence_diagrams/Anticipation/StaleDataIgnore_Dark.png?text=DarkMode"/>
</figure>

If `StaleDataHandling` is set to `StaleDataHandling.Reanticipate`, stale data will be treated the same way as any other server data updates. The value will be rolled back, `ShouldReanticipate` will be set to true, and the `OnReanticipate` event will be fired (see below). In typical client prediction systems, and generally involves replaying the player's input from the time of the incoming data to now, which results in re-performing the switch to red.

> 

Reviewer 2: @jabbacakes
Not anything to hold this PR up, but we should see if the samples team has any bandwidth to create a series of companion examples to this documentation as it could be confusing to try to replicate the code required for this. Kitty already provided a generlized sample that has a lot of this done, so it might not be too time consuming to make something (based on the sample Kitty provided) that would walk users through this entire page of documentation.

### Global OnReanticipate

In addition to the `NetworkBehaviour`'s `OnReanticipate` method, `NetworkManager` also has a callback that can be subscribed to for global reanticipation. This is useful if you need to run your reanticipation in a more global way, such as if you need to run it step-wise (say, anticipating one frame at a time) and need all objects to complete one step before any of them begin the second one. This callback receives the same `lastRoundTripTime` value as the `NetworkBehaviour` method, and is called **after** all of the `NetworkBehaviour` methods have been called.

:::note

If you want to implement a full client-side prediction model in your game, the global OnReanticipate callback is likely the ideal place to incorporate your rollback and replay logic. The details of implementing this, however, are left as an exercise for the user; implementing a full, production-ready prediction loop is a complex topic and recommended for advanced users only.

:::

> ## Smoothing

Reviewer: ```suggestion
## Smoothing for Corrections
```

> Much of the time, when you perform an anticipation action, your anticipation will be correct and the server will update the state to the same value as you did. Occasionally, though, the server will have a different result, since latency can result in the server state changing before it receives your command in ways that affect the result. When this happens, by default, the value will be snapped to the new server value.

Reviewer: I'd maybe try to add a link to this section at the top when explaining the basics of anticipation. I was reading the first paragraphs and I kept thinking "what about when there's desyncs?"

In some cases, this may result in an undesirable player experience, causing things that normally move smoothly to become choppy. When this happens, you can smooth out the result by using the `Smooth()` function on both `AnticipatedNetworkVariable` and `AnticipatedNetworkTransform`.

> `Smooth()` takes a starting position and an ending position (which would usually be the prevoius anticipated value and either the new anticipated value or the current authoritatvie value) and a duration over which to perform the smoothing. For `AnticipatedNetworkVariable`, because the type is generic, it also requires a delegate to use to perform the smoothing operation (which takes a from value, a to value, and a percent value, and in many cases may just be `Mathf.Lerp`).

Reviewer: ```suggestion
`Smooth()` takes a starting position and an ending position (which would usually be the previous anticipated value and either the new anticipated value or the current authoritative value) and a duration over which to perform the smoothing. For `AnticipatedNetworkVariable`, because the type is generic, it also requires a delegate to use to perform the smoothing operation (which takes a from value, a to value, and a percent value, and in many cases may just be `Mathf.Lerp`).
```

> Because the first parameter to `Smooth()` is usually the previous anticipated value, and by the time the `OnReanticipate` function is called, the anticipated value has been overwritten with the authoritative value, you can use `PreviousAnticipatedValue` on `AnticipatedNetworkVariable` and `PreviousAnticipatedState` on `AnticipatedNetworkTransform` to access the value you most recently passed into `Anticipate()` or its related transform methods. Note, however, that if you call `Anticipate()` or its related transform methods prior to doing the smoothing, they will update the previous anticipated value, so you may need to store the previous anticipated value in a local variable before doing your reanticipation.

Reviewer: Hopefully that's something Amy can help with, that doc has very long sentences :P

### Server-Side Smoothing

Even though these concepts are mostly for client-side use, there are cases where you might want to perform server-side smoothing as well. For example, when running in host mode, the host player will see the movement of other players. Due to latency and jitter, input updates from remote clients may not arrive every frame, which may result in choppy movement of an AnticipatedNetworkTransform, for example.

> To address this, you can also use the `Smooth()` function at any time on the server side. The server does not do anticipation, so it will not get any `OnReanticipate()` events, but you can, for example, call `Smooth()` each time you receive an input from the client in order to smooth the motion between the previous position and the new one.

Reviewer: Should that really be handled by Netcode? Couldn't users interpolate their own authoritative data or just use Physics interpolation for example?

Author: On the client side, yes, though it's such a common thing that I figured it would be useful to give users a method for it.

On the server side, smoothing is still used for interpolation, and it was a huge pain to implement it for AnticipatedNetworkTransform. They'd have a hard time implementing it themselves.

> One important distinction between client-side smoothing and server-side smoothing for `AnticipatedNetworkTransform`, though, is that client-side smoothing smooths the actual motion of the transform, while server-side smoothing only smooths the visual depiction of the motion. Which is to say, `AnticipatedState` is updated, but `transform` is not, so all game logic and collision detection will be done based on the actual position, and the rendering of the object will be slightly behind the actual position of the object.

Reviewer: As a user, how would I access the authoritative value vs the visual value? What APIs do I use? For example, if my camera should track the interpolated value, how do I access it?

Reviewer: From going back above, I'm assuming that's gameObject.transform for interpolated? And this.GetComponent<AnticipatedNetworkTransform>().authoritativeState for authoritative value?
